File: app/src/main/java/com/blog/demo/image/paint/PaintPathActivity.java
package com.blog.demo.image.paint;

import android.app.Activity;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.CornerPathEffect;
import android.graphics.DashPathEffect;
import android.graphics.DiscretePathEffect;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PathDashPathEffect;
import android.graphics.PathEffect;
import android.os.Bundle;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;

public class PaintPathActivity extends Activity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(new PaintPathView(this));
    }

    class PaintPathView extends View {
        private Path mPath = new Path();
        private Paint mPaint = new Paint();
        private int mPhase = 0;

        public PaintPathView(Context context) {
            this(context, null);
        }

        public PaintPathView(Context context, @Nullable AttributeSet attrs) {
            super(context, attrs);

            mPath.moveTo(50, 50);
            for (int index = 1; index < 20; index++) {
                mPath.lineTo(50 + 30 * index, 50 + (float) (60 * Math.random()));
            }

            mPaint.setAntiAlias(true);
            mPaint.setColor(Color.BLACK);
            mPaint.setStyle(Paint.Style.STROKE);
            mPaint.setStrokeWidth(2);
        }

        @Override
        protected void onDraw(Canvas canvas) {
            Path path = new Path();
            path.addCircle(0, 0, 3, Path.Direction.CW);

            PathEffect[] effects = new PathEffect[]{null,
                    new CornerPathEffect(25),
                    new DashPathEffect(new float[]{20, 5, 10, 5}, mPhase),
                    new PathDashPathEffect(path, 12, mPhase, PathDashPathEffect.Style.ROTATE),
                    new DiscretePathEffect(3.0f, 5.0f),
                    new DiscretePathEffect(5.0f, 3.0f)
            };

            drawPath(canvas, 0, effects[0]);

            drawPath(canvas, 80, effects[1]);

            drawPath(canvas, 160, effects[2]);

            drawPath(canvas, 240, effects[3]);

            drawPath(canvas, 320, effects[4]);
            drawPath(canvas, 400, effects[5]);

            mPhase++;
            invalidate();
        }

        private void drawPath(Canvas canvas, int dy, PathEffect effect) {
            canvas.save();

            canvas.translate(0, dy);
            mPaint.setPathEffect(effect);
            canvas.drawPath(mPath, mPaint);

            canvas.restore();
        }

    }


}

File: app/src/main/java/com/blog/demo/image/paint/PaintPathActivity.kt
package com.blog.demo.image.paint

import android.app.Activity
import android.content.Context
import android.graphics.*
import android.os.Bundle
import android.util.AttributeSet
import android.view.View

class PaintPathActivity : Activity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(PaintPathView(this))
    }

    private inner class PaintPathView(context: Context?, attrs: AttributeSet? = null) : View(context, attrs) {
        private val mPath = Path()
        private val mPaint = Paint()
        private var mPhase = 0

        init {
            mPath.moveTo(50f, 50f)
            for (index in 1..19) {
                mPath.lineTo((50 + 30 * index).toFloat(), 50 + (60 * Math.random()).toFloat())
            }
            mPaint.isAntiAlias = true
            mPaint.color = Color.BLACK
            mPaint.style = Paint.Style.STROKE
            mPaint.strokeWidth = 2f
        }

        override fun onDraw(canvas: Canvas) {
            val path = Path()
            path.addCircle(0f, 0f, 3f, Path.Direction.CW)

            val effects = arrayOf(CornerPathEffect(25f),
                DashPathEffect(floatArrayOf(20f, 5f, 10f, 5f), mPhase.toFloat()),
                PathDashPathEffect(path, 12f, mPhase.toFloat(), PathDashPathEffect.Style.ROTATE),
                DiscretePathEffect(3.0f, 5.0f),
                DiscretePathEffect(5.0f, 3.0f))
            drawPath(canvas, 0, null)
            drawPath(canvas, 80, effects[0])
            drawPath(canvas, 160, effects[1])
            drawPath(canvas, 240, effects[2])
            drawPath(canvas, 320, effects[3])
            drawPath(canvas, 400, effects[4])

            mPhase++
            mPhase %= 1000
            invalidate()
        }

        private fun drawPath(canvas: Canvas, dy: Int, effect: PathEffect?) {
            canvas.save()

            canvas.translate(0f, dy.toFloat())
            mPaint.pathEffect = effect
            canvas.drawPath(mPath, mPaint)

            canvas.restore()
        }

    }

}